README.md
# ipfs Cookbook

This cookbook installs ipfs and starts it as a daemon

## Requirements

### Platforms

This cookbook is tested on Ubuntu 16.06, 14.04 and Debian 8 using Test Kitchen.
It currently only supports 64bit platforms

### Chef

- Chef 12.5 or later (we are providing a
  [https://docs.chef.io/custom_resources.html](Custom Resource) to configure
  IPFS

### Cookbook dependencies

- `ark` to download and uncompress the Go IPFS package

## Usage

### ipfs::default

Just include `ipfs` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[ipfs]"
  ]
}
```

## Attributes

- `node.['ipfs']['version']` - the Go IPFS version to download from the official
site (64bit)
- `node['ipfs']['checksum']` - the SHA256 checksum for the package
- `node['ipfs']['config']['swarm']['addr_filter']` - the network ranges to not
connect to. This will stop platforms like Hetzner to block your server
(https://github.com/ipfs/go-ipfs/issues/1226)

## Resources

`ipfs_config` sets the config. Supports hashes, arrays, booleans and strings.
Does not change anything if the config already has that value, and restarts
the server automatically

```ruby
ipfs_config "Gateway.Writable" do
  value "true"
 end
```

## Running the specs and integrations tests

Install the latest [Chef DK](https://downloads.chef.io/chefdk).

```
chef exec delivery local all # Run the linting check, syntax check and unit tests
kitchen verify # Run the integration tests for Ubuntu 16.04, 18.06 and Debian 9
```

## License and Authors

Authors: Kosmos Developers mail@kosmos.org
Copyright: 2018, [Kosmos Developers](https://kredits.kosmos.org/)

```
Unless otherwise noted, all files are released under the MIT license, possible
exceptions will contain licensing information in them.

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
# ipfs CHANGELOG

## 0.1.0
- [gregkare] - Initial release of ipfs


spec/unit/recipes/default_spec.rb
#
# Cookbook:: ipfs
# Spec:: default
#
# Copyright 2018, Kosmos
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'ipfs::default' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on Ubuntu 18.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '18.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/cluster_service_spec.rb
#
# Cookbook:: ipfs
# Spec:: default
#
# Copyright 2018, Kosmos
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'ipfs::cluster_service' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'enables the service' do
      expect(chef_run).to enable_service('ipfs-cluster')
    end
  end

  context 'When all attributes are default, on Ubuntu 18.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '18.04')
      runner.converge(described_recipe)
    end

    it 'enables the service' do
      expect(chef_run).to enable_service('ipfs-cluster')
    end
  end
end


test/integration/default/serverspec/ipfs_spec.rb
require 'serverspec'

# Required by serverspec
set :backend, :exec

describe 'IPFS' do
  # It is in the PATH
  describe command('which ipfs') do
    its(:exit_status) { should eq 0 }
  end

  it 'is listening on port 4001' do
    expect(port(4001)).to be_listening
  end

  it 'is listening on port 8080 (gateway)' do
    expect(port(8080)).to be_listening
  end

  it 'has a running service of ipfs' do
    expect(service('ipfs')).to be_running
    expect(service('ipfs')).to be_enabled
  end
end


recipes/_user.rb
#
# Cookbook Name:: ipfs
# Recipe:: _user
#
# The MIT License (MIT)
#
# Copyright:: 2018, Kosmos Developers
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

group 'ipfs' do
  gid 4737
end

user 'ipfs' do
  comment 'ipfs'
  uid 4737
  gid 4737
  home '/home/ipfs'
  manage_home true
end


recipes/cluster.rb
#
# Cookbook Name:: ipfs
# Recipe:: cluster
#
# The MIT License (MIT)
#
# Copyright:: 2018, Kosmos Developers
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

include_recipe 'ipfs::_user'

version = node['ipfs']['cluster']['version']

ark 'ipfs-cluster-service' do
  url "https://dist.ipfs.io/ipfs-cluster-service/v#{version}/ipfs-cluster-service_v#{version}_linux-amd64.tar.gz"
  has_binaries ['ipfs-cluster-service']
end

ark 'ipfs-cluster-ctl' do
  url "https://dist.ipfs.io/ipfs-cluster-ctl/v#{version}/ipfs-cluster-ctl_v#{version}_linux-amd64.tar.gz"
  has_binaries ['ipfs-cluster-ctl']
end

credentials = data_bag_item('credentials', 'ipfs_cluster')

execute 'ipfs-cluster-service init' do
  user 'ipfs'
  environment 'CLUSTER_SECRET' => credentials['secret'],
              'IPFS_CLUSTER_PATH' => '/home/ipfs/.ipfs-cluster'
  not_if { File.exist? '/home/ipfs/.ipfs-cluster/service.json' }
end


resources/config.rb
require 'json'
require 'mixlib/shellout'

property :key, String, name_property: true
property :value, [String, Hash, Array, TrueClass, FalseClass], required: true

action :create do
  include_recipe 'ipfs'

  execute "ipfs config --json #{new_resource.key} '#{JSON.generate(new_resource.value)}'" do
    environment 'IPFS_PATH' => '/home/ipfs/.ipfs'
    user 'ipfs'
    not_if do
      cmd = Mixlib::ShellOut.new('ipfs', 'config', new_resource.key,
                                 user: 'ipfs',
                                 env: { 'IPFS_PATH' => '/home/ipfs/.ipfs' })
      cmd.run_command
      begin
        JSON.parse(cmd.stdout) == JSON.parse(JSON.generate(new_resource.value))
      rescue JSON::ParserError
        cmd.stdout.include?(new_resource.value.to_s)
      end
    end
    notifies :restart, 'service[ipfs]', :delayed
  end
end


recipes/cluster_service.rb
#
# Cookbook Name:: ipfs
# Recipe:: cluster_service
#
# Copyright 2018, Kosmos
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if platform?('ubuntu') && node['platform_version'].to_f < 15.04 ||
   platform?('debian') && node['platform_version'].to_f < 8
  template 'ipfs.initd-cluster.service.erb' do
    path '/etc/init.d/ipfs-cluster'
    source 'ipfs-cluster.initd.service.erb'
    owner 'root'
    group 'root'
    mode '0750'
    notifies :restart, 'service[ipfs-cluster]', :delayed
  end

  service 'ipfs-cluster' do
    provider Chef::Provider::Service::Init::Debian
    action [:enable]
    supports start: true, stop: true, restart: true, reload: false, status: true
  end

else
  execute 'systemctl daemon-reload' do
    command 'systemctl daemon-reload'
    action :nothing
  end

  template 'ipfs-cluster.systemd.service.erb' do
    path '/lib/systemd/system/ipfs-cluster.service'
    source 'ipfs-cluster.systemd.service.erb'
    owner 'root'
    group 'root'
    mode '0644'
    notifies :run, 'execute[systemctl daemon-reload]', :delayed
    notifies :restart, 'service[ipfs-cluster]', :delayed
  end

  service 'ipfs-cluster' do
    provider Chef::Provider::Service::Systemd
    action [:enable]
  end
end


recipes/cluster_bootstrap.rb
#
# Cookbook Name:: ipfs
# Recipe:: cluster_bootstrap
#
# Copyright 2018, Kosmos
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'ipfs::_user'

execute "ipfs-cluster-service daemon --bootstrap #{peer}" do
  user 'ipfs'
  environment 'IPFS_CLUSTER_PATH' => '/home/ipfs/.ipfs-cluster'
  not_if { File.exist? '/home/ipfs/.ipfs-cluster/service.json' }
end


recipes/default.rb
#
# Cookbook Name:: ipfs
# Recipe:: default
#
# Copyright 2017-2018, Kosmos
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'ipfs::_user'

version = node['ipfs']['version']

ark 'ipfs' do
  url "https://dist.ipfs.io/go-ipfs/v#{version}/go-ipfs_v#{version}_linux-amd64.tar.gz"
  checksum node['ipfs']['checksum']
  has_binaries ['ipfs']
end

execute 'ipfs init --empty-repo' do
  environment 'IPFS_PATH' => '/home/ipfs/.ipfs'
  user 'ipfs'
  not_if { File.directory? '/home/ipfs/.ipfs' }
end

if platform?('ubuntu') && node['platform_version'].to_f < 15.04 ||
   platform?('debian') && node['platform_version'].to_f < 8
  template 'ipfs.initd.service.erb' do
    path '/etc/init.d/ipfs'
    source 'ipfs.initd.service.erb'
    owner 'root'
    group 'root'
    mode '0750'
    notifies :restart, 'service[ipfs]', :delayed
  end

  service 'ipfs' do
    provider Chef::Provider::Service::Init::Debian
    action [:enable]
    supports start: true, stop: true, restart: true, reload: false, status: true
  end

else
  execute 'systemctl daemon-reload' do
    command 'systemctl daemon-reload'
    action :nothing
  end

  template 'ipfs.systemd.service.erb' do
    path '/lib/systemd/system/ipfs.service'
    source 'ipfs.systemd.service.erb'
    owner 'root'
    group 'root'
    mode '0644'
    notifies :run, 'execute[systemctl daemon-reload]', :delayed
    notifies :restart, 'service[ipfs]', :delayed
  end

  service 'ipfs' do
    provider Chef::Provider::Service::Systemd
    action [:enable]
  end
end

# Configure ipfs to not contact local network addresses
ipfs_config 'Swarm.AddrFilters' do
  value node['ipfs']['config']['swarm']['addr_filter']
end
